### PR TITLE
Tests for task's API

### PR DESCRIPTION
**Pull request by psteinb**
As discussed in #19 this PR isolated the unit test for `two_moons` to serve as a starting point establishing tests for all tasks.

### CODE REVIEW

**Comment by jan-matthis** [comment]
This is great, cheers! 

What do you think about parametrizing those tests such that they run for all tasks, i.e., something along the lines of:
```
@pytest.mark.parametrize(
    "task_name",
    [
        (task_name,) for task_name in sbibm.get_available_tasks()
    ],
)
```

**Comment by psteinb** [comment]
I think that this makes a lot of sense. I can add that if you want. 
I suggest to be a bit more selective though in terms of what is tested, in order not to extend the runtime of the tests. One way to go would perhaps be, to do these automated tests in `/repo/tests/task` as they mostly test the interface of the tests which is settled in `/repo/sbibm/tasks/task.py`. 
For tests potentially running longer, we could mirror the same structure as in `/repo/sbibm/tasks/ and allow for specialized tests (e.g. see the `.*demo.*` test cases in this PR) in the `/repo/tests/tasks/two_moons/test_special.py`. I think this keeps the balance between a "god test suite" and more specialized tests if they are needed. And I'd hope, it makes contributing easier.

**Comment by psteinb** [comment]
@jan-matthis done for now. I am happy to adapt #18 if this can be merged earlier.

File: tests/tasks/test_task_benchmarks.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+#     assert outputs.shape
+#     assert outputs.shape[0] > 0
+#     assert logprob_truep == None
+
+#     predictive_samples = sim(outputs)
+#     x_o = task.get_observation(7)
+#     value = median_distance(predictive_samples, x_o)
+
+#     assert value > 0
```
Comment (jan-matthis) [inline tests/tasks/test_task_benchmarks.py:75]: This should go out if we are not using it

File: tests/tasks/test_task_benchmarks.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+import pytest
+import torch
+
+from sbibm import get_available_tasks, get_task
+from sbibm.algorithms.sbi.snpe import run
+from sbibm.metrics.ppc import median_distance
+
+# maybe use the pyro facilities
+torch.manual_seed(47)
```
Comment (jan-matthis) [inline tests/tasks/test_task_benchmarks.py:11]: Yes, let's use `set_rng_seed` from `pyro.utils` instead which will also seed `numpy` and `random`

File: tests/tasks/test_task_interface.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
+import re
+
+import pytest
+import torch
+
+from sbibm import get_available_tasks, get_task
+
+# maybe use the pyro facilities
```
Comment (jan-matthis) [inline tests/tasks/test_task_interface.py:8]: Yes (see above)

**Review by jan-matthis — COMMENTED** [review]
Great, I think this provides a good foundation for testing existing and future tasks. Thanks a lot for your work on this, much appreciated! 

I only left small comments

File: tests/tasks/test_task_benchmarks.py (hunk 1 above)
Comment (psteinb) [reply tests/tasks/test_task_benchmarks.py]: done with 26d92f5

File: tests/tasks/test_task_benchmarks.py (hunk 2 above)
Comment (psteinb) [reply tests/tasks/test_task_benchmarks.py]: done with 9e91818

File: tests/tasks/test_task_interface.py (hunk 3 above)
Comment (psteinb) [reply tests/tasks/test_task_interface.py]: done with 9e91818

**Comment by psteinb** [comment]
Thanks for the review. I hope I implemented those comments alright.

File: tests/tasks/two_moons/test_task.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import pytest
+import torch
+
+from sbibm.tasks.two_moons.task import TwoMoons
+
+torch.manual_seed(47)
```
Comment (jan-matthis) [inline tests/tasks/two_moons/test_task.py:6]: Could you swap this one for ` pyro.util.set_rng_seed(47)` as well?

**Review by jan-matthis — COMMENTED** [review]
Except for a single comment this looks good to go from my side, cheers!

File: tests/tasks/two_moons/test_task.py (hunk 4 above)
Comment (psteinb) [reply tests/tasks/two_moons/test_task.py:6]: I could, I would take the liberty to do that everywhere then:
```
$ git --no-pager grep 'torch.manual_seed'
sbibm/tasks/bernoulli_glm/task.py:            torch.manual_seed(observation_seed)
sbibm/tasks/slcp/task.py:        torch.manual_seed(42)
sbibm/tasks/task.py:            torch.manual_seed(observation_seed)
tests/tasks/two_moons/test_task.py:torch.manual_seed(47)
```

Comment (psteinb) [reply tests/tasks/two_moons/test_task.py]: done [`9ffc0f6` (#29)](https://github.com/sbi-benchmark/sbibm/pull/29/commits/9ffc0f6d7e7e808fac43c067510561d3f0cc1e12)

Comment (jan-matthis) [reply tests/tasks/two_moons/test_task.py]: Thanks for checking where else this appears. I would rather restrict the changes to `tests/` in this PR though -- changing the other occurrences would change seeding for regenerating observations which I'd like to avoid

Comment (psteinb) [reply tests/tasks/two_moons/test_task.py]: :cry: ok, I picked up some git-foo this week with respect to rebasing. Let me see if I can ditch this commit that overdid the pyro-seeding.

Comment (jan-matthis) [reply tests/tasks/two_moons/test_task.py]: Should be pretty straighforward -- all you need is `git rebase -i main` (on the feature branch) and then drop the last one :)

Comment (psteinb) [reply tests/tasks/two_moons/test_task.py]: done with [`abc4a85` (#29)](https://github.com/sbi-benchmark/sbibm/pull/29/commits/abc4a852b00641546cb6591cc72e34ee4da81b61)

**Comment by jan-matthis** [comment]
Cheers!
